### PR TITLE
Emit deprecation warning for duplicate files with default duplicate strategy

### PR DESCRIPTION
**Pull request by wolfs**
Instead of failing.

See https://github.com/gradle/gradle/pull/10143 which introduced the change to fail the build on duplicates.

### CODE REVIEW

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ArchiveIntegrationTest.groovy [hunk 1]
```diff
@@ -687,7 +687,7 @@ class ArchiveIntegrationTest extends AbstractIntegrationSpec {
         "test.tar"  | "tarTree" | "createTar"
     }
 
-    def 'ensure duplicates not included in tar by default'() {
+    def 'emit deprecation warning when duplicates are included in tar for default duplicates strategy'() {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ArchiveIntegrationTest.groovy]: So `ArchiveIntegrationTest` is about TAR, and `ZipIntegrationTest` is about ZIP? Looks like something we should consolidate.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy [hunk 2]
```diff
@@ -27,7 +27,7 @@ import java.nio.charset.Charset
 @TestReproducibleArchives
 class ZipIntegrationTest extends AbstractIntegrationSpec {
 
-    def 'ensure duplicates not included by default'() {
+    def 'deprecation warning on duplicates with default strategy'() {
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy]: I guess this should have a similar name in both test classes.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy [hunk 3]
```diff
@@ -40,10 +40,9 @@ class ZipIntegrationTest extends AbstractIntegrationSpec {
             }
         '''
 
-        when:
-        fails 'zip'
-        then:
-        failure.assertHasCause('Encountered duplicate path "file1.txt" during copy operation configured with DuplicatesStrategy.FAIL')
+        expect:
+        executer.expectDeprecationWarning('Encountered duplicate path during copy operation configured with default duplicates strategy. Allowing duplicates with the default duplicates strategy has been deprecated. This is scheduled to be removed in Gradle 7.0. Duplicate path: "file1.txt". Explicitly set the duplicates strategy to \'DuplicatesStrategy.INCLUDE\' if you want to allow duplicate paths.')
```
Comment (lptr) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy]: > Allowing duplicates with the default duplicates strategy has been deprecated.

This sounds contrived. I don't think it adds much to the (already long) message, so WDYT about removing it?

Comment (lptr) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy]: I'd move mentioning the duplicate entry towards the beginning of the message.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java [hunk 4]
```diff
@@ -218,13 +220,19 @@ public ContentFilterable expand(Map<String, ?> properties) {
     @Override
     public void setDuplicatesStrategy(DuplicatesStrategy strategy) {
         this.duplicatesStrategy = strategy;
+        this.defaultDuplicatesStrategy = strategy == DuplicatesStrategy.INHERIT;
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java]: Why do we store this instead of calculating it when needed?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/FileCopyDetailsInternal.java [hunk 5]
```diff
@@ -22,4 +22,7 @@
 
     boolean isIncludeEmptyDirs();
 
+    default boolean isDefaultDuplicatesStrategy() {
+        throw new UnsupportedOperationException();
+    }
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/FileCopyDetailsInternal.java]: There are only two implementations, and it's not something users should extend. I think the `default` implementation (especially one so punishing) is not warranted here.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy [hunk 6]
```diff
@@ -2352,6 +2352,7 @@ Found the following transforms:
             ${declareTransform(transformImplementation)}
 
             task resolve(type: Copy) {
+                duplicatesStrategy = 'INCLUDE'
```
Comment (lptr) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy]: Why is this needed?

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 7]
```diff
@@ -77,18 +77,14 @@ The related APIs in `StartParameter` are now deprecated.
...
 
-=== Potential breaking changes
-
-==== Android Gradle Plugin 3.3 and earlier is not supported anymore
-
-Gradle 6.0 supports Android Gradle Plugin versions 3.4 and later.
+==== Copy or archive duplicate paths without explicitly specifying a duplicates strategy has been deprecated
 
-==== Archive tasks fail on duplicate files
+Copy and archive tasks default to the `INCLUDE` duplicates strategy, allowing the same path to be copied multiple times to the destination or to exist multiple times in an archive.
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
Archive tasks `Tar` and `Zip` by default allow multiple entries for the same path to exist in the created archive.
These https://github.com/gradle/gradle/issues/9990["grossly invalid zip files" can trigger zip bomb detection].
To prevent this from happening accidentally encountering duplicates while creating an archive now produces a deprecation message, and will fail the build starting with Gradle 7.0.

Note that copy tasks also happily copy multiple sources with the same relative path to the destination directory.
That behavior has also been deprecated.
```

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc [hunk 8]
```diff
@@ -77,18 +77,14 @@ The related APIs in `StartParameter` are now deprecated.
...
 
-Until now archive tasks defaulted to the `INCLUDE` duplicates strategy, allowing the same path to exist multiple times in an archive.
+Trying to copy or archive duplicate paths without explicitly specifying a duplicates strategy is now deprecated.
+This is to prevent user from accidentally creating https://github.com/gradle/gradle/issues/9990["grossly invalid zip files" that trigger zip bomb detection].
 
-In Gradle 6.0 we are switching to `FAIL`, prohibiting duplicate files in archives.
-If you still want to allow them, you can be specify that explicitly:
+If you still want to allow them, you can specify that explicitly:
```
Comment (lptr) [inline subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]:
```suggestion
If you want to allow duplicates, you can specify that explicitly:
```

I think "still" carries a kind of condescending tone here without adding anything, so I'd remove it.
Replacing "them" with specific noun to make it easier to understand it without having to read through the previous sentences.

File: subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppApplicationPublishingIntegrationTest.groovy [hunk 9]
```diff
@@ -38,7 +38,6 @@ class CppApplicationPublishingIntegrationTest extends AbstractCppPublishingInteg
     def consumer = file("consumer").createDir()
 
     def setup() {
-        when:
```
Comment (lptr) [inline subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppApplicationPublishingIntegrationTest.groovy]: Hah!

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy (hunk 3 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy]: The deprecation message needs to contain something with `... has been deprecated`. Moreover, I wanted to have the file which caused the duplication as context, so the given deprecations count as the same. Do you agree with those constraints? Do you have a suggestion to make the message better having those constraints in mind?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java (hunk 4 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DefaultFileCopyDetails.java]: When the default duplicates strategy is used, then this class is initialized with the duplicates strategy `INCLUDE`. Only when the duplicates strategy is set on the `FileCopyDetails` later on we need to update whether the default strategy is still used.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy (hunk 6 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy]: Since we sometimes resolve two `jar`s with the same name and copy them into the same directory.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java [hunk 10]
```diff
@@ -38,32 +38,30 @@ public DuplicateHandlingCopyActionDecorator(CopyAction delegate) {
...
-                                    LOGGER.warn("Encountered duplicate path \"{}\" during copy operation configured with DuplicatesStrategy.WARN", details.getRelativePath());
-                                }
-                            }
-                        }
-
-                        action.processFile(details);
+                if (!visitedFiles.add(details.getRelativePath())) {
+                    if (details.isDefaultDuplicatesStrategy()) {
+                        DeprecationLogger.nagUserWithDeprecatedIndirectUserCodeCause(
+                            "Encountered duplicate path during copy operation configured with default duplicates strategy. Allowing duplicates with the default duplicates strategy",
```
[the remark this answers is not in the document]
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java]: How about `Copying or archiving duplicate paths with the default duplicates strategy has been deprecated. This is scheduled to be removed in Gradle 7.0.`?

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc (hunk 8 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: Done.

File: subprojects/docs/src/docs/userguide/upgrading_version_5.adoc (hunk 7 above)
Comment (wolfs) [reply subprojects/docs/src/docs/userguide/upgrading_version_5.adoc]: Done.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/FileCopyDetailsInternal.java (hunk 5 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/FileCopyDetailsInternal.java]: I removed the default method

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy (hunk 2 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ZipIntegrationTest.groovy]: Both tests are now in the same class with a `where` class.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ArchiveIntegrationTest.groovy (hunk 1 above)
Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/bundling/ArchiveIntegrationTest.groovy]: `ArchiveIntegrationTest` is for tar and zip. I moved the duplicates tests to `ArchiveIntegrationTest` so we only need to maintain them in one place.

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java]: 👍

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy (hunk 6 above)
Comment (lptr) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy]: Isn't that something we should fix?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/NormalizingCopyActionDecorator.java [hunk 11]
```diff
@@ -53,31 +53,26 @@ public NormalizingCopyActionDecorator(CopyAction delegate, Chmod chmod) {
 
     @Override
     public WorkResult execute(final CopyActionProcessingStream stream) {
-        final Set<RelativePath> visitedDirs = new HashSet<RelativePath>();
+        final Set<RelativePath> visitedDirs = new HashSet<>();
         final ListMultimap<RelativePath, FileCopyDetailsInternal> pendingDirs = ArrayListMultimap.create();
 
-        WorkResult result = delegate.execute(new CopyActionProcessingStream() {
+        return delegate.execute(new CopyActionProcessingStream() {
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/copy/NormalizingCopyActionDecorator.java]: Nit: make this into a lambda (by moving the private method to the enclosing class)?

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy (hunk 6 above)
Comment (wolfs) [reply subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy]: I don't think so. The point of the task is to resolve the configurations involved, the tests don't care where the artifacts are copied to.

File: subprojects/core/src/main/java/org/gradle/api/internal/file/copy/NormalizingCopyActionDecorator.java (hunk 11 above)
Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/copy/NormalizingCopyActionDecorator.java]: Done.
